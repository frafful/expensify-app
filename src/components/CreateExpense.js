import React from 'react';

const CreateExpense = () => (
  <div>
    Create expense component
  </div>
);

export default CreateExpense;