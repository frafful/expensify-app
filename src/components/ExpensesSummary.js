import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/selectExpensesTotal';
import filteredExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
  
  const expenseWord = expensesCount > 1 ? 'expenses' : 'expense';
  return (
    <div>
      { 
        expensesCount > 0 && 
          <h1>
            Viewing {expensesCount} {expenseWord} totalling { numeral((expensesTotal / 100)).format('$0,0.00') }
          </h1> 
      }
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expensesCount: filteredExpenses(state.expenses, filteredExpenses).length,
    expensesTotal: selectExpensesTotal(filteredExpenses(state.expenses, state.filters))
  };
};

export default connect(mapStateToProps)(ExpensesSummary);


