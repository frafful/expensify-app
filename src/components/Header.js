import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (  
  <header>
    <h1>Expensify</h1>
    <nav>
      <ul>
        <li><NavLink exact to="/" activeClassName="is-active">Home</NavLink></li>
        <li><NavLink exact to="/create" activeClassName="is-active">Add Expense</NavLink></li>
        <li><NavLink exact to="/help" activeClassName="is-active">Help</NavLink></li>
      </ul>
    </nav>
  </header>
);

export default Header;