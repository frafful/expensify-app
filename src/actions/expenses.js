import database from '../firebase/firebase';
import { EditExpense } from '../components/EditExpense';

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '', 
      amount = 0, 
      createdAt = 0 
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    });
  };
};

export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
        const expensesData = [];
        
        snapshot.forEach((childSnapshot) => {
          expensesData.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        
        dispatch(setExpenses(expensesData));
    });
  }
};

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then( () => {
      dispatch(removeExpense({ id }));
    });
  };
};

export const startEditExpense = ({ id }, updatedExpenseValues) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updatedExpenseValues).then(() => {
      dispatch(editExpense(id, updatedExpenseValues));
    });
  };
}