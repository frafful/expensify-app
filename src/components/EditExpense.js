import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
  
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }
  
  render() {
    return (
      <div>
        <ExpenseForm 
          onSubmit={this.onSubmit}
          expense={this.props.expense}
        />
        <button onClick={this.onClick}>
          Remove
        </button>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch, props) => (
  {
    startEditExpense: (id, expense) => (dispatch(startEditExpense({ id }, expense))),
    startRemoveExpense: (id) => (dispatch(startRemoveExpense(id)))
  }
);

const mapStateToProps = (state, props) => {
  const expense = state.expenses.find((exp) => props.match.params.id === exp.id);
  return {
    expense: expense
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);