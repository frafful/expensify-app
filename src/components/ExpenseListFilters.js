import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment'
import { setStartDate, setEndDate, setTextFilter, sortByDate, sortByAmount } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      calendarFocusType: null
    }
  }

  onDatesChange = ({startDate, endDate}) => {
    startDate && this.props.setStartDate(startDate);
    endDate && this.props.setEndDate(endDate);
  };

  onCalendarFocusChange = (focusedInput) => { 
    this.setState((state) => ({
      calendarFocusType: focusedInput
    }));
  };

  onSortChange = (e) => {
    e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  render() {
    return (
      <div>
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
        <select 
          value={ this.props.filters.sortBy }
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate} 
          endDate={this.props.filters.endDate} 
          onDatesChange={this.onDatesChange} 
          focusedInput={this.state.calendarFocusType}
          onFocusChange={this.onCalendarFocusChange} 
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
  };
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);