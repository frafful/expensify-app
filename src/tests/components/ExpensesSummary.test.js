import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import expenses from '../fixtures/expenses';
import { ExpensesSummary } from '../../components/ExpensesSummary';


test('should not render summary if no expenses exists', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={0} expensesTotal={0}/>);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render Expense totals component correctly', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={100000}/>);
  expect(toJSON(wrapper)).toMatchSnapshot();
});