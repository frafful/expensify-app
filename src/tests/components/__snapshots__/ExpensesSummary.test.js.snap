// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render summary if no expenses exists 1`] = `<div />`;

exports[`should render Expense totals component correctly 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totalling 
    $1,000.00
  </h1>
</div>
`;
